### PR TITLE
KAFKA-10017: fix flaky EosBetaUpgradeIntegrationTest

### PR DESCRIPTION
**Pull request by ableegoldman**
The current failures we're seeing with this test are due to faulty assumptions that it makes and not any real bug in eos-beta (at least, from what I've seen so far).

The test relies on tightly controlling the commits, which it does by setting the commit interval to MAX_VALUE and manually requesting commits on the context. In two phases, the test assumes that any pending data will be committed after a rebalance. But we actually take care to avoid unnecessary commits -- with eos-alpha, we only commit tasks that are revoked while in eos-beta we must commit all tasks if any are revoked, but only if the revoked tasks themselves need a commit.

The failure we see occurs when we try to verify the committed data after a second client is started and the group rebalances. The already-running client has to give up two tasks to the newly started client, but those tasks may not need to be committed in which case none of the tasks would be. So we still have an open transaction on the partitions where we try to read committed data.

We can use a punctuator to force a commit on the running client

### CODE REVIEW

**Comment by abbccdda** [comment]
retest this please

**Comment by ableegoldman** [comment]
EosBetaUpgradeIntegrationTest still failed a few times on the PR builds, but it seems to be actual flakiness and not a new bug in eos-beta. I'll try and tighten up the test a bit more

`org.apache.kafka.streams.errors.InvalidStateStoreException: Cannot get state store store because the stream thread is PARTITIONS_ASSIGNED, not RUNNING`
and
`java.lang.AssertionError: Condition not met within timeout 60000. Client did not reach 10 stable assignments on time, numStableAssignments was 11`

**Review by guozhangwang — COMMENTED** [review]
The explanation makes sense to me. Thanks @ableegoldman for digging into this!

LMK if you want to have a quick sync regarding the potential places that needs to be tightened.

**Comment by guozhangwang** [comment]
test this please

**Comment by ableegoldman** [comment]
Looks like it only failed one time across all three builds in the last run. But I pushed some fixes for the remaining flakiness:
1. Use the IntegrationTestUtils `#getStore` to retry on InvalidStateStoreException
2. Relax the assignment listener check: the way we currently make sure to reach a stable assignment is to save the current number of stable assignments before a rebalance-triggering event (eg 2nd client is stopped) and then wait for the number of stable assignments to increase by exactly one. In some cases the test failed because we actually got to more than the expected number of stable assignments. I'm not exactly sure why, but I don't think it's necessarily a problem and we can probably just relax the condition to make sure we had at least one more stable assignment

Can you kick off the tests again @guozhangwang ? I'm interested to see if the fix for 2. is enough or if whatever is causing the extra rebalance will surface as a test failure in another way

**Comment by ijuma** [comment]
ok to test

**Comment by ijuma** [comment]
retest this please

**Comment by vvcephei** [comment]
Test this please

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Would it speed up the test to choose a smaller number here?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:172]: There should never be multiple requests, right? If there were, a second request might arrive between 168 and 169, violating the desired property. In that case, we should grab a lock instead. As long as there's only one requesting thread, and it always waits for the commit right after requesting, then we should be good.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:871]: There's going to be a separate punctuator per task, right? Does the test account for this?

**Review by vvcephei — COMMENTED** [review]
Woah, thanks for the fix, @ableegoldman !

**Comment by abbccdda** [comment]
Seems still got 3 failures, while 1/3 green

**Comment by abbccdda** [comment]
retest this please

**Comment by ableegoldman** [comment]
> Seems still got 3 failures, while 1/3 green

@abbccdda Did the EosBetaUpgradeIntegrationTest fail 3 times? Did you get the link or stacktrace before kicking off tests again?
edit: Nevermind, I assume you meant other flaky tests

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:871]: Sort of; with eos-beta when you need to commit one task, you need to commit all of them, so requesting a commit at all should be sufficient. It's kind of a subtle point, but I think it's actually preferable to just request a commit on one task so this test can also verify that we commit all tasks together

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:172]: This is only meant to be used when there's a single thread (and single instance running). I just wanted to put in a quick fix for now--I'll leave a comment to warn our future selves that we'll have to tighten this up if we want to extend the test or use it elsewhere with multiple threads/instances

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: I suppose it would. Does 100ms seem reasonable?

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Yeah, that's what I was thinking

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:172]: sounds good

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:871]: Ok, I was thinking more along the lines that we have just one requestCommit instance while we would have multiple punctuator instances all using it. It looked like a bug, but it seems like it'll be fine.

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Retest this please

**Comment by vvcephei** [comment]
Retest this please

**Review by vvcephei — APPROVED** [review]
LGTM. Thanks, @ableegoldman !

**Comment by vvcephei** [comment]
When we merge this, we should cherry-pick it to 2.6 and remove the ignore.

**Comment by ijuma** [comment]
It would be good to run this in a loop locally 100 times (or something like that) to check that it's truly stable.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:871]: In this test we would have multiple punctuator indeed but they would be executed by a single thread sequentially so that's fine.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:1276]: Makes sense to me.

**Review by guozhangwang — APPROVED** [review]
My understanding is that we can only merge this PR after merging https://github.com/apache/kafka/pull/8962 since both contributes to the integration test failure right? Anyways, the tighten-up all looks good to me.

**Comment by ableegoldman** [comment]
@guozhangwang I don't _think_ we need [8962](https://github.com/apache/kafka/pull/8962) to be merged first

**Review by andrewchoi5 — APPROVED** [review]

**Comment by guozhangwang** [comment]
Merged to 2.6 as well and re-enabled the integration test.

**Pull request merged**
